x/ibc/ante.go

```diff
@@ -1,51 +1,70 @@
 package ibc
 
-// // TODO: Should extract timeout msgs too
-// func ExtractMsgPackets(msgs []sdk.Msg) (res []MsgPacket, abort bool) {
-// 	res = make([]MsgPacket, 0, len(msgs))
-// 	for _, msg := range msgs {
-// 		msgp, ok := msg.(MsgPacket)
-// 		if ok {
-// 			res = append(res, msgp)
-// 		}
-// 	}
-
-// 	if len(res) >= 2 {
-// 		first := res[0]
-// 		for _, msg := range res[1:] {
-// 			if len(msg.ChannelID) != 0 && msg.ChannelID != first.ChannelID {
-// 				return res, true
-// 			}
-// 			msg.ChannelID = first.ChannelID
-// 		}
-// 	}
-
-// 	return
-// }
-
-// func VerifyMsgPackets(ctx sdk.Context, channel channel.Manager, msgs []MsgPacket) error {
-// 	for _, msg := range msgs {
-// 		err := channel.Receive(ctx, msg.Proofs, msg.Height, msg.ReceiverPort(), msg.ChannelID, msg.Packet)
-// 		if err != nil {
-// 			return err
-// 		}
-// 	}
-
-// 	return nil
-// }
-
-// func NewAnteDecorator(channel channel.Manager) sdk.AnteDecorator {
-// 	return func(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
-// 		msgs, abort := ExtractMsgPackets(tx.GetMsgs())
-// 		if abort {
-// 			return ctx, host.ErrInvalidPacket
-// 		}
-
-// 		err := VerifyMsgPackets(ctx, channel, msgs)
-// 		if err != nil {
-// 			return ctx, sdkerrors.Wrap(host.ErrInvalidPacket, err.Error())
-// 		}
-
-// 		return next(ctx, tx, simulate)
-// 	}
-// }
+import (
+	"errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	client "github.com/cosmos/cosmos-sdk/x/ibc/02-client"
+	channel "github.com/cosmos/cosmos-sdk/x/ibc/04-channel"
+)
+
+type ProofVerificationDecorator struct {
+	clientKeeper  client.Keeper
+	channelKeeper channel.Keeper
+}
+
+func (pvr ProofVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	var flag bool
+	var portID, channelID string
+
+	for _, msg := range tx.GetMsgs() {
+		var err error
+		switch msg := msg.(type) {
+		case client.MsgUpdateClient:
+			err = pvr.clientKeeper.UpdateClient(ctx, msg.ClientID, msg.Header)
+		case channel.MsgPacket:
+			err = pvr.channelKeeper.RecvPacket(ctx, msg.Packet, msg.Proof, msg.ProofHeight)
+			if flag {
+				if portID != msg.DestinationPort || channelID != msg.DestinationChannel {
+					return ctx, errors.New("Transaction cannot include IBC packets from different channels")
+				}
+			} else {
+				portID = msg.DestinationPort
+				channelID = msg.DestinationChannel
+				flag = true
+			}
+			pvr.channelKeeper.SetPacketAcknowledgement(ctx, msg.DestinationPort, msg.DestinationChannel, msg.Sequence, []byte{})
```

Comment:
This should be returned from `RecvPacket`, not always `[]byte{}`.

Reply:
This is just for convenience - each app handler will override the acknowledgement using `WriteAcknowledgement` only if it is not empty bytes, and they can omit the function if the ack is empty bytes because it is already set inside the ante.

Reply:
Why not have `recvPacket` return the acknowledgement? (and write it if there is no error)